### DISCUSSION TITLE
"Apparent Greyscale" transform

### DISCUSSION THREAD

**Discussion by dg1727** [libvips/libvips#1996] [Ideas]
I wonder if the VIPS team might be interested in adding the following algorithm for converting colour to greyscale:  
http://resources.mpi-inf.mpg.de/ApparentGreyscale/
I didn't find mention of it in the VIPS documentation, including the output of the `vips list classes` commandline (as of vips-7.42.3).

**Comment by jcupitt** [comment]
Hi, in fact by coincidence vips got a better greyscale a couple of days ago:

https://github.com/jcupitt/libvips/issues/266

I tried their test image with vips master and got this:

![vips-greyscale](https://cloud.githubusercontent.com/assets/580843/7229585/54bea3e6-e75d-11e4-8a1a-4e2504f57c20.png)

Which (I think) actually looks better. They have too much contrast in the foreground (in my opinion).

**Comment by dg1727** [comment]
I think "contrast in the foreground" is what they are aiming for.  Many people would say that the red flower in the original stands out against the blue and green around it, so it's right for it to have high contrast in the greyscale as well.  

It seems like you're pushing for a soon release of vips 8.0, and I wouldn't want that to be delayed, but maybe this other greyscale method can be considered after that release?  

Or if you don't run the project that way, then you can close this suggestion, and if others make the same request later, then maybe this feature will be added.

**Comment by jcupitt** [comment]
Sure, let's leave this open and tagged as "enhancement". Their algorithm does look interesting.

**Reply by daniellovera** [reply]
> Sure, let's leave this open and tagged as "enhancement". Their algorithm does look interesting.

Hi @jcupitt . Was the apparent greyscale algorithm ever implemented as an enhancement? The original issue was marked as completed in Feb 2021, though I searched the code and documents and couldn't find a reference to this greyscale transform being added.

I switched to using libvips for greyscale transform: 
```python
vips.Image.new_from_array(img, interpretation='auto').colourspace('b-w').numpy()
``` 
from using openCV
```python
cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
``` 

And, if I'm understanding correctly, it turns out that openCV uses this for greyscale conversion in cvtColor: `Y = 0.299*R + 0.587*G + 0.114*B` whereas libvips is using [.2 .7 .1](https://github.com/libvips/libvips/blob/ec432a57968b8c9c240cdef0343437eb59571929/libvips/colour/colour.c#L136)

Interestingly, this caused some material changes to the thresholding operations I was performing on the converted-to-greyscale images, so it led me to this thread. The apparent greyscale transform, in my testing with images converted with the GIMP plug-in, performed the best for my use cases, so I figured I'd check if it was added to libvips before I hack together some python.

**Reply by jcupitt** [reply]
Hi @daniellovera, it's still in the backlog of interesting things, but hasn't bubbled to the top.

You're right, we should change libvips to use that ratio as well. But I think that's probably not what you're seeing -- libvips is doing rgb -> y in linear space (ie. gamma correct conversion), whereas (I think?) opencv is just merging the rgb bands.

I've not tested this idea, which is silly of me, but gamma correct conversion generally has a much more dramatic effect than these relatively small coefficient changes.

**Reply by jcupitt** [reply]
I changed the libvips coefficients and credited you (I hope that's OK): https://github.com/libvips/libvips/commit/2d676076b5b42ba6bf177e13111da7e8ff5c9380

**Reply by daniellovera** [reply]
Perfectly okay, glad to have contributed.

That said, this was more a squirrel finding a nut than deep understanding and analysis, so I'm curious as to why libvips should change the coefficients to use that ratio. Is it because [.2 .7 .1] is generally accurate enough for greyscale conversion and thus was used to date, but [0.299 0.587 0.114] is the more precise coefficient set for going from scRGB 2 BW, which is now part of every colour route to get to BW?

I'm intrigued since my understanding is that it's [0.2126729, 0.7151522, 0.0721750] for sRGB to greyscale, calculated after gamma correction is applied to linearize the RGB values.

I may be off here and confusing concepts, though I think the reason I'm seeing this is because I happened to be analyzing red text on a blue background that lined up with the Helmholtz-Kohlrausch effect. The OpenCV coefficients essentially increased contrast in the greyscale image by 50%; I don't think gamma correction comes into play here since it would net out? I guess what I'm saying is that I do think using apparent greyscale as the transform to BW could have a dramatic effect on computer vision tasks that rely on thresholding, such as OCR and scene/object detection, and if there's a way I could accelerate bubbling apparent greyscale towards the top of the list I would be thrilled to do so.

**Reply by jcupitt** [reply]
Ah I just read the wiki page too quickly. I've changed it again to 0.2126. 0.7152, 0.0722: https://github.com/libvips/libvips/commit/2ad93b3f013e073879832aaf63a3ff14f88eea2c Thanks for pointing out my brain fart.

Gamma has a surprisingly strong effect -- any time you do arithmetic on pixel values you should really be in linear space or you'll see distortions. In these images:

![vips-greyscale](https://cloud.githubusercontent.com/assets/580843/7229585/54bea3e6-e75d-11e4-8a1a-4e2504f57c20.png)

It's original, gimp greyscale, apparent greyscale, libvips greyscale. The gimp-vips differences are just down to gamma.

I agree apparent greyscale could be a useful thing for many tasks, it's a question of finding the time to do it. I'm on "sabbatical" right now and working on another project, but if someone would like to volunteer to do the work, I'd be delighted to advise, if that'd be useful.

**Reply by daniellovera** [reply]
I'm willing to volunteer and give it a go.

Question, is CIELUV colourspace implemented in libvips? I see references to it regarding TIFF files and this [thread](https://github.com/libvips/libvips/issues/1506), though I do not see it elsewhere. Does CIELUV go by a different name?

I ask because the apparent greyscale [paper](https://resources.mpi-inf.mpg.de/ApparentGreyscale/Perceptual_Greyscale.pdf) uses CIELUV for the step 1, and then CIELAB for step 2. My initial thought was the colourspace conversion needed would be Lab2AppG (example name) with a detour to CIELUV to calculate Nayatani's chromatic lightness model, though I haven't yet explored if Nayatani's model can be implemented in CIELAB colourspace.
